File: C/第4章 栈/4.2 最小栈/min_stack.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

typedef struct {
    int *data;
    int top;
} Stack;

typedef struct {
    Stack *stack;
    Stack *minStack;
} MinStack;

int min(int num1, int num2) {
    return num1 < num2 ? num1 : num2;
}

MinStack* init(int size) {
    MinStack *minStack = (MinStack *)malloc(sizeof(MinStack));

    minStack->stack = (Stack *)malloc(sizeof(Stack));
    minStack->stack->data = (int *)malloc(sizeof(int) * size);
    minStack->stack->top = -1;

    minStack->minStack = (Stack *)malloc(sizeof(Stack));
    minStack->minStack->data = (int *)malloc(sizeof(int) * size);
    minStack->minStack->data[0] = INT_MAX;
    minStack->minStack->top = 0;

    return minStack;
}

void push(MinStack *minStack, int data) {
    minStack->stack->data[++minStack->stack->top] = data;
    minStack->minStack->data[++minStack->minStack->top] = min(data, minStack->minStack->data[minStack->minStack->top]);
}

void pop(MinStack *minStack) {
    minStack->stack->top--;
    minStack->minStack->top--;
}

int peek(MinStack *minStack) {
    return minStack->stack->data[minStack->stack->top];
}

int getMin(MinStack *minStack) {
    return minStack->minStack->data[minStack->minStack->top];
}

void deleteMinStack(MinStack *minStack) {
    free(minStack->stack->data);
    free(minStack->stack);
    free(minStack->minStack->data);
    free(minStack->minStack);
    free(minStack);
}

int main() {
    MinStack *minStack = init(5);

    push(minStack, -2);
    push(minStack, 0);
    push(minStack, -3);

    printf("%d\n", getMin(minStack));       // -3
    pop(minStack);
    printf("%d\n", peek(minStack));         // 0
    printf("%d\n", getMin(minStack));       // -2

    deleteMinStack(minStack);
    return 0;
}

File: Python/第4章 栈/4.2 最小栈/min_stack.py
import math

class MinStack:
    def __init__(self):
        self.stack = []
        self.min_stack = [math.inf]
    
    def push(self, data):
        self.stack.append(data)
        self.min_stack.append(min(data, self.min_stack[-1]))
    
    def pop(self):
        self.stack.pop()
        self.min_stack.pop()
    
    def peek(self):
        return self.stack[-1]
    
    def get_min(self):
        return self.min_stack[-1]

def main():
    min_stack = MinStack()

    min_stack.push(-2)
    min_stack.push(0)
    min_stack.push(-3)

    print(min_stack.get_min())      # -3
    min_stack.pop()
    print(min_stack.peek())         # 0
    print(min_stack.get_min())      # -2

if __name__ == "__main__":
    main()